### DISCUSSION TITLE
FPDualReal in Fluid Properties UO

### DISCUSSION THREAD

**Discussion by joe61vette** [idaholab/moose#17061] [Q&A Modules: General] [answered]
Hi:

Very basic question.  Trying to code a UO for a fluid property that would use AD system.  In the header file I have:

class ConstantFluidProperties : public SinglePhaseFluidProperties
...
virtual FPDualReal rho_from_p_T(const FPDualReal & pressure, const FPDualReal & temperature) const override;

and in the corresponding C file:

FPDualReal
ConstantFluidProperties::rho_from_p_T(const FPDualReal & pressure, const FPDualReal & temperature) const
{
...
}

I get a compile error:

ConstantFluidProperties.h:23:23: error: 'rho_from_p_T' marked 'override' but does not override any member functions
   virtual FPDualReal rho_from_p_T(const FPDualReal & pressure, const FPDualReal & temperature) const override;

It seems like it should be overriding propfuncAD.  Is there something obvious that I am doing wrong? 

Thanks,
   Joe K

**Comment by GiudGiud** [comment]
Hi Joe

I think it's because propfuncAD is using DualReal for its arguments and return types. Going to check now. 

Guillaume

**Reply by GiudGiud** [reply]
Also I think you can define your functions with non-AD Real types and propfuncAD will take care of defining the AD versions.

**Reply by GiudGiud** [reply]
I think defining with non-AD Real types and have the propfuncAD template do the conversion is the way to go here.
@lindsayad is more comfortable than me with AD maybe he will have better advice

**Reply by joe61vette** [reply]
Thanks Guillaume:  I wondered about using DualReal but tried to follow BrineFluidProperties as a template.  It inherits from MultiComponentFluidProperties which does contain a new macro for propfuncAD to account for the extra independent variable.  But that macro (if I understand the coding) also uses DualReal but the the header and C files have:

  FPDualReal rho_from_p_T_X(const FPDualReal & pressure,
                            const FPDualReal & temperature,
                            const FPDualReal & xnacl) const;
and

FPDualReal
BrineFluidProperties::rho_from_p_T_X(const FPDualReal & pressure,
                                     const FPDualReal & temperature,
                                     const FPDualReal & xnacl) const

Also, I think FPDualReal is just a specialization of DualReal:

typedef DualNumber<Real, DNDerivativeSize<5>> FPDualReal;

  Joe

**Reply by GiudGiud** [reply]
Yes the issue was not because of FPDualReal.

The routine in BrineFluidProperties you pasted is rho_from_p_T**_X**, which is not provided by propfuncAD I think, so it's not marked with override in the header. You could try just deleting the override keyword for your declaration, but then your routine could conflict with the one defined by propfuncAD.

**Reply by cpgr** [reply]
Hi Joe,

I added that `FPDualReal` typedef to compute some tricky derivatives in `BrineFluidProperties` using AD. It was much faster to limit the number of derivatives (5 derivates was heaps faster than allowing 50). I think I also use it in `WaterFluidProperties` for the same reason. You might get some clues how they can be used in conjunction with `propfunc` macros looking there, eg https://github.com/idaholab/moose/blob/6437ec83fe011ae67c7b71b4210b426db6ca320b/modules/fluid_properties/src/userobjects/Water97FluidProperties.C#L842

**Reply by joe61vette** [reply]
Thanks Chris: 

From Water97FluidProperties it seems that to use FPDualReal, one has to forget about inheriting from propfuncAD and instead write a new fn.  For example, there is:

FPDualReal
Water97FluidProperties::T_from_p_h_ad(const FPDualReal & pressure, const FPDualReal & enthalpy) const

where the "_ad" has been added but is not in the macro.  Using FPDualReal seems to make a lot of sense but it appears that using it is outside of the normal guidance.  As this is a new app, I would like to try and follow the Moose guidelines as much as possible.

If I only use DualReal, all is well.  If I use FPDualReal and omit the "override" specifier, it compiles but I get the following run time error:

*** ERROR ***
The following error occurred in the object "const_props", of type "ConstantFluidProperties".

const_props: virtual void SinglePhaseFluidProperties::rho_from_p_T(libMesh::Real, libMesh::Real, libMesh::Real &, libMesh::Real &, libMesh::Real &) const derivatives not implemented.

Is it best to just go ahead and use DualReal (accepting performance penalty) but follow the fluid property system guidelines?

  Joe

**Reply by cpgr** [reply]
In your app, you probably want to call `rho_from_p_T(DualReal p, DualReal T)` to use the AD capability. But the way the `propfunc` macros work in the fluid properties classes is that you add non-AD versions, and then that macro automatically creates the AD version. This means that we can use those fluid properties with `Real` variables, and also `DualReal` variables, so we have a lot of flexibility.

So in your class, you want to add:
```
Real rho_from_p_T(Real p, Real T)
{
  return rho = f(p, T) // for some function f
}

void rho_from_p_T(Real p, Real T, Real & rho, Real & rho_dp, Real & rho_dT)
{
  rho = f(p, T)
  rho_dp = df_dp // whatever these derivatives are
  rho_dT = df_dT
}
```
and then you get `DualReal rho_from_p_T(DualReal & p, DualReal & T)` for free without coding them! This is what the macros add for you.

Where I have used `FPDualReal` is in place of hand-coding the derivates in `rho_from_p_T(Real p, Real T, Real & rho, Real & rho_dp, Real & rho_dT)`. Perhaps a better example than I gave earlier is https://github.com/idaholab/moose/blob/6437ec83fe011ae67c7b71b4210b426db6ca320b/modules/fluid_properties/src/userobjects/Water97FluidProperties.C#L1683-L1697

So I use the faster (less derivatives) AD form to compute the derivatives, and then get the full AD version through the macro automatically. Does that explain it?

**Reply by joe61vette** [reply]
THANKS!  I finally got it.  So, the "approved" way is to follow the normal

Real rho_from_p_T(Real p, Real T)

type of coding and let the macro handle it.  I tested it and it does work for my case where rho is an AD Material Property and p & T are adCoupledValues.  One only implements the extra FPDualReal coding (as in your example) if the normal way is simply too slow.  Which for Water/Steam properties would be the case.

I'll mark this as answered.  Thanks again,
   Joe

**Reply by joe61vette** [reply]
Well, I would have marked it as answered if I could figure out how.  Joe

**Reply by lindsayad** [reply]
Unfortunately you can only mark the beginning of the thread as the answer. 

FWIW with the new sparse AD default configuration of MOOSE, `DualReal` should not be so slow any more because it keeps track of the number of derivative entries automatically and consequently dual number operations only happen over the sparsity entries (which can be a small number, less than 5 for example, in these fluid property contexts) and not over the entire underlying container size of 50.

**Reply by cpgr** [reply]
Thanks for the info @lindsayad. I'll do some comparisons when I get the time and swap those instances of `FPDualReal` to normal `DualReal` if there isn't much of a difference.

**Reply by lindsayad** [reply]
You would definitely pay a memory penalty though because derivative vector is still statically sized

**Comment by joe61vette** [comment, marked as answer]
Just want to thank everyone for the help.  It is great that I can simply use the old style property fns and have them magically return the value with derivatives.  Also, using the AD system as Chris did to get derivatives for properties that return them will save me time later.
